File: content/_code-samples/use-tickets/py/use-tickets.py
from xrpl.clients import JsonRpcClient
from xrpl.models.transactions import TicketCreate, AccountSet
from xrpl.transaction import safe_sign_and_submit_transaction
from xrpl.wallet import Wallet, generate_faucet_wallet
from xrpl.models.requests.account_objects import AccountObjects, AccountObjectType

# Connect to a testnet node
JSON_RPC_URL = "https://s.altnet.rippletest.net:51234/"
client = JsonRpcClient(JSON_RPC_URL)

# Generate a wallet and request faucet
test_wallet = generate_faucet_wallet(client=client)
myAddr = test_wallet.classic_address

# Construct a TicketCreate transaction, 2 ticket created for future use
tx = TicketCreate(
    account=myAddr,
    ticket_count=2
)

# Sign transaction locally and submit
my_tx_payment_signed = safe_sign_and_submit_transaction(transaction=tx, wallet=test_wallet, client=client)

# Get a Ticket Sequence
get_ticket_sequence = AccountObjects(
    account=myAddr,
    type=AccountObjectType.TICKET
)

response = client.request(get_ticket_sequence)

# Since we created 2 Tickets previously, you're able to choose which Ticket you're going to use
ticket1_sequence = response.result["account_objects"][0]["TicketSequence"]
ticket2_sequence = response.result["account_objects"][1]["TicketSequence"]

print(f"Ticket 1: {ticket1_sequence}\n"
      f"Ticket 2: {ticket2_sequence}")

ticket_sequence = int(input("Pick a ticket sequence to use for your next transaction: "))

# Construct Transaction using a Ticket
tx_1 = AccountSet(
    account=myAddr,
    fee="10",
    sequence=0,
    last_ledger_sequence=None,
    ticket_sequence=ticket_sequence
)

# Send transaction (w/ Ticket)
tx_result = safe_sign_and_submit_transaction(transaction=tx_1, client=client, wallet=test_wallet)
result = tx_result.result["engine_result"]

print(f"Account: {myAddr}")
if result == "tesSUCCESS":
    print("Transaction successful!")
elif result == "terPRE_TICKET":
    print("The provided Ticket Sequence does not exist in the ledger")
elif result == "tefMAX_LEDGER":
    print("Transaction failed to achieve consensus")
elif result == "tefPAST_SEQ":
    print("The sequence number is lower than the current sequence number of the account")
elif result == "unknown":
    print("Transaction status unknown")
else:
    print(f"Transaction failed with code {result}")

File: content/_code-samples/use-tickets/js/use-tickets.js
// Dependencies for Node.js.
// In browsers, use a <script> tag instead.
if (typeof module !== "undefined") {
  // Use var here because const/let are block-scoped to the if statement.
  var xrpl = require('xrpl')
}

// Example credentials
const wallet = xrpl.Wallet.fromSeed("sn3nxiW7v8KXzPzAqzyHXbSSKNuN9")

// Connect to Devnet (since that's where tickets are available)
async function main() {
  const client = new xrpl.Client("wss://s.devnet.rippletest.net:51233")
  await client.connect()

  // Get credentials from the Testnet Faucet -----------------------------------
  console.log("Getting a wallet from the faucet...")
  const {wallet, balance} = await client.fundWallet()

  // Check Sequence Number -----------------------------------------------------
  const account_info = await client.request({
    "command": "account_info",
    "account": wallet.address
  })
  let current_sequence = account_info.result.account_data.Sequence

  // Prepare and Sign TicketCreate ---------------------------------------------
  const prepared = await client.autofill({
    "TransactionType": "TicketCreate",
    "Account": wallet.address,
    "TicketCount": 10,
    "Sequence": current_sequence
  })
  const signed = wallet.sign(prepared)
  console.log(`Prepared TicketCreate transaction ${signed.hash}`)

  // Submit TicketCreate -------------------------------------------------------
  const tx = await client.submitAndWait(signed.tx_blob)
  console.log(tx)

  // Wait for Validation -------------------------------------------------------
  // submitAndWait() handles this automatically, but it can take 4-7s.

  // Check Available Tickets ---------------------------------------------------
  let response = await client.request({
    "command": "account_objects",
    "account": wallet.address,
    "type": "ticket"
  })
  console.log("Available Tickets:", response.result.account_objects)

  // Choose an arbitrary Ticket to use
  use_ticket = response.result.account_objects[0].TicketSequence

  // Prepare and Sign Ticketed Transaction -------------------------------------
  const prepared_t = await client.autofill({
    "TransactionType": "AccountSet",
    "Account": wallet.address,
    "TicketSequence": use_ticket,
    "LastLedgerSequence": null, // Never expire this transaction.
    "Sequence": 0
  })
  const signed_t = wallet.sign(prepared_t)
  console.log(`Prepared ticketed transaction ${signed_t.hash}`)

  // Submit Ticketed Transaction -----------------------------------------------
  const tx_t = await client.submitAndWait(signed_t.tx_blob)
  console.log(tx_t)

  // Wait for Validation (again) -----------------------------------------------

  // Disconnect when done (If you omit this, Node.js won't end the process)
  await client.disconnect()
}

main()
